### PR TITLE
Cache cufft plans

### PR DESCRIPTION
**Pull request by ssnl**
#8120 
Empirical speed up when running our test 10 times is `9.57s => 7.90s` with cache enabled.

### CODE REVIEW

File: aten/src/ATen/Context.h [hunk 1]
```diff
@@ -111,6 +111,8 @@ class AT_API Context {
   void setBenchmarkCuDNN(bool);
   bool deterministicCuDNN() const;
   void setDeterministicCuDNN(bool);
+  bool cachePlanCuFFT() const;
+  void setCachePlanCuFFT(bool);
```
Comment (ezyang) [inline aten/src/ATen/Context.h]: Is there a reason to turn caching on/off on a global purpose, as opposed to just having caching on all the time or providing a caching/non-caching function?

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -5446,6 +5461,11 @@ def parse_kwargs(desc):
     For CPU tensors, this method is currently only available with MKL. Check
     :func:`torch.backends.mkl.is_available` to check if MKL is installed.
 
+.. note::
+    For CUDA tensors, setting ``torch.backends.cuda.cufft_cache_plan = True``
+    turns on cuFFT plan caching, and thus speed up repeatedly running FFT
+    methods on tensors of same shape and strides with same figurations.
```
Comment (ezyang) [inline torch/_torch_docs.py]: configurations

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -5209,6 +5209,11 @@ def parse_kwargs(desc):
     For CPU tensors, this method is currently only available with MKL. Check
     :func:`torch.backends.mkl.is_available` to check if MKL is installed.
 
+.. note::
+    For CUDA tensors, setting ``torch.backends.cuda.cufft_cache_plan = True``
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_torch_docs.py]: Actually, I was mistaken. We do always *check* the cache (even when benchmarking is turned off), but it looks like we don't store results in the cache unless benchmarking is on.

File: aten/src/ATen/third_party/ParamsMap.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
+#pragma once
```
Comment (ezyang) [inline aten/src/ATen/third_party/ParamsMap.h]: Mmm, does this really belong in `third_party`?

File: aten/src/ATen/native/cuda/CuFFTUtils.h [hunk 5]
```diff
@@ -66,19 +70,261 @@ static inline void CUFFT_CHECK(cufftResult error)
...
-public:
-  explicit CufftHandle() {
-    CUFFT_CHECK(cufftCreate(&raw_plan));
+
+// This POD struct is used to let us easily compute hashes of the
+// parameters
+// It will be the **key** to the plan cache.
+struct CuFFTParams
+{
+  at::ScalarType scalar_type;
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/CuFFTUtils.h]: Consider suffixing all of the members with a `_`? Will make it easier to distinguish parameter updates.

**Review by ezyang — APPROVED** [review]
I'm going to approve this, but I am a little concerned about the discoverability of the caching mechanism. Should we perhaps figure out some way to turn the caching on by default, e.g., setup some sort of eviction policy in the cache?

**Comment by SsnL** [comment]
Now using an LRU cache.

docs are not updated yet. I want to check that CI passes first.

**Comment by SsnL** [comment]
:D CUDA hook worked pretty well. Will update the docs soon.

File: aten/src/ATen/function_wrapper.py [hunk 6]
```diff
@@ -967,14 +970,14 @@ def find_formal(formal_name, formals):
...
         is_factory_method = find_formal('TensorOptions', formals)
-        is_deprecated_factory_method = \
-            formals[0]['dynamic_type'] == 'Type' and option['return_type'] == 'Tensor' and option['deprecated']
+        is_deprecated_factory_method = len(formals) > 0 and \
+            formals[0]['dynamic_type'] == 'Type' and \
+            option['return_type'] == 'Tensor' and option['deprecated']
         needs_native_definition = not is_deprecated_factory_method
 
         has_dispatch = dispatch_tensor or dispatch_type
-        is_namespace_function = is_function and (has_dispatch or is_factory_method)
```
Comment (SsnL) [inline aten/src/ATen/function_wrapper.py]: I've verified that all other functions w/ `is_function = True` have `(has_dispatch or is_factory_method) = True` too. So I think relaxing the check here is fine.

**Review by orionr — COMMENTED** [review]
Is the third_party/onnx change actually required for this PR?

**Comment by SsnL** [comment]
@orionr reverted that change

**Review by ezyang — APPROVED** [review]

**Comment by SsnL** [comment]
ci failure looks unrelated

**Pull request merged**